File: web/js/trading.js
/**
 * trading.js — Quote validation, trade execution, and market management.
 */

import { Quote, Trade, Phase } from './state.js';

export class TradingError extends Error {
  constructor(message) {
    super(message);
    this.name = 'TradingError';
  }
}

/**
 * Validate and return a Quote. Spread must be exactly 2.
 */
export function validateQuote(bid, ask) {
  if (ask !== bid + 2) {
    throw new TradingError(`Spread must be exactly 2 (bid=${bid}, ask=${ask}). Try ask = bid + 2.`);
  }
  return new Quote(bid, ask);
}

/**
 * Set a player's two-way market. Validates spread and records it.
 */
export function setMarket(state, playerId, bid, ask) {
  if (!state.playerIds.includes(playerId)) {
    throw new TradingError(`Unknown player: ${playerId}`);
  }
  if (state.phase === Phase.SETTLE) {
    throw new TradingError('Cannot quote during settlement.');
  }
  const quote = validateQuote(bid, ask);
  state.markets[playerId] = quote;
  return quote;
}

/**
 * Execute a trade where aggressor hits market_maker's quote.
 */
export function executeTrade(state, aggressor, marketMaker, direction) {
  if (aggressor === marketMaker) {
    throw new TradingError('Cannot trade with yourself.');
  }
  if (state.phase === Phase.SETTLE) {
    throw new TradingError('Cannot trade during settlement.');
  }

  const quote = state.markets[marketMaker];
  if (!quote) {
    throw new TradingError(`${marketMaker} has not posted a market yet.`);
  }

  let trade;
  if (direction === 'buy') {
    trade = new Trade(aggressor, marketMaker, quote.ask, state.phase);
  } else if (direction === 'sell') {
    trade = new Trade(marketMaker, aggressor, quote.bid, state.phase);
  } else {
    throw new TradingError(`Direction must be 'buy' or 'sell', got '${direction}'.`);
  }

  state.trades.push(trade);
  return trade;
}

/**
 * Return the current market for a player, or null if not yet posted.
 */
export function getQuote(state, playerId) {
  return state.markets[playerId] || null;
}

/**
 * Return { nBuys, nSells, avgBuy, avgSell } for a player.
 */
export function tradeSummary(state, playerId) {
  const buys = state.trades.filter(t => t.buyer === playerId).map(t => t.price);
  const sells = state.trades.filter(t => t.seller === playerId).map(t => t.price);
  const avgBuy = buys.length > 0 ? buys.reduce((a, b) => a + b, 0) / buys.length : 0.0;
  const avgSell = sells.length > 0 ? sells.reduce((a, b) => a + b, 0) / sells.length : 0.0;
  return { nBuys: buys.length, nSells: sells.length, avgBuy, avgSell };
}

/**
 * Execute a trade directly at a specific price, without looking up
 * a standing market quote. Used by the V2 ask-for-a-price flow.
 */
export function executeTradeDirect(state, buyer, seller, price) {
  if (buyer === seller) {
    throw new TradingError('Cannot trade with yourself.');
  }
  if (state.phase === Phase.SETTLE) {
    throw new TradingError('Cannot trade during settlement.');
  }
  if (!state.playerIds.includes(buyer)) {
    throw new TradingError(`Unknown player: ${buyer}`);
  }
  if (!state.playerIds.includes(seller)) {
    throw new TradingError(`Unknown player: ${seller}`);
  }

  const trade = new Trade(buyer, seller, price, state.phase);
  state.trades.push(trade);
  return trade;
}

File: cli/game/trading.py
"""
trading.py — Quote validation, trade execution, and market management.
"""

from __future__ import annotations

from typing import Optional, Tuple

from game.state import GameState, Quote, Trade, Phase


class TradingError(Exception):
    """Raised when an invalid trade or quote is attempted."""


def validate_quote(bid: int, ask: int) -> Quote:
    """
    Validate and return a Quote. Spread must be exactly 2.
    Raises TradingError on invalid input.
    """
    if ask != bid + 2:
        raise TradingError(f"Spread must be exactly 2 (bid={bid}, ask={ask}). Try ask = bid + 2.")
    return Quote(bid=bid, ask=ask)


def set_market(state: GameState, player_id: str, bid: int, ask: int) -> Quote:
    """
    Set a player's two-way market. Validates spread and records it.
    Returns the Quote.
    """
    if player_id not in state.player_ids:
        raise TradingError(f"Unknown player: {player_id}")
    if state.phase == Phase.SETTLE:
        raise TradingError("Cannot quote during settlement.")
    quote = validate_quote(bid, ask)
    state.markets[player_id] = quote
    return quote


def execute_trade(
    state: GameState,
    aggressor: str,
    market_maker: str,
    direction: str,  # "buy" or "sell"
) -> Trade:
    """
    Execute a trade where `aggressor` hits `market_maker`'s quote.

    direction="buy"  → aggressor buys at market_maker's ask price
    direction="sell" → aggressor sells at market_maker's bid price

    Records the trade and returns the Trade object.
    """
    if aggressor == market_maker:
        raise TradingError("Cannot trade with yourself.")
    if state.phase == Phase.SETTLE:
        raise TradingError("Cannot trade during settlement.")

    quote = state.markets.get(market_maker)
    if quote is None:
        raise TradingError(f"{market_maker} has not posted a market yet.")

    if direction == "buy":
        price = quote.ask
        trade = Trade(buyer=aggressor, seller=market_maker, price=price, phase=state.phase)
    elif direction == "sell":
        price = quote.bid
        trade = Trade(buyer=market_maker, seller=aggressor, price=price, phase=state.phase)
    else:
        raise TradingError(f"Direction must be 'buy' or 'sell', got '{direction}'.")

    state.trades.append(trade)
    return trade


def get_quote(state: GameState, player_id: str) -> Optional[Quote]:
    """Return the current market for a player, or None if not yet posted."""
    return state.markets.get(player_id)


def trade_summary(state: GameState, player_id: str) -> Tuple[int, int, float, float]:
    """
    Return (n_buys, n_sells, avg_buy_price, avg_sell_price) for a player.
    Averages are 0.0 if no trades of that type.
    """
    buys = [t.price for t in state.trades if t.buyer == player_id]
    sells = [t.price for t in state.trades if t.seller == player_id]
    avg_buy = sum(buys) / len(buys) if buys else 0.0
    avg_sell = sum(sells) / len(sells) if sells else 0.0
    return len(buys), len(sells), avg_buy, avg_sell


# ── V2: Direct trade execution (no standing markets required) ─────────────────

def execute_trade_direct(
    state: GameState,
    buyer: str,
    seller: str,
    price: int,
) -> Trade:
    """
    Execute a trade directly at a specific price, without looking up
    a standing market quote. Used by the V2 ask-for-a-price flow.

    buyer:  player who buys (gains if final_total > price)
    seller: player who sells (gains if final_total < price)
    price:  agreed transaction price
    """
    if buyer == seller:
        raise TradingError("Cannot trade with yourself.")
    if state.phase == Phase.SETTLE:
        raise TradingError("Cannot trade during settlement.")
    if buyer not in state.player_ids:
        raise TradingError(f"Unknown player: {buyer}")
    if seller not in state.player_ids:
        raise TradingError(f"Unknown player: {seller}")

    trade = Trade(buyer=buyer, seller=seller, price=price, phase=state.phase)
    state.trades.append(trade)
    return trade
